javascript/README.md

```diff
@@ -285,3 +291,113 @@ const { a, ...noA } = copy; // noA => { b: 2, c: 3 }
 ```
 
 [⇧ top](#javascript-guide)
+
+---
+
+## Destructuring
+
+### 4.1 Object Destructuring
+
+Use [object destructuring][Object Destructuring] when accessing and using multiple properties of an object.
+
+> Why? Destructuring saves you from creating temporary references for those properties.
+
+#### Examples
+
+🚫 Nope. 🚫
+
+```js
+function getFullName(user) {
+  const firstName = user.firstName;
+  const lastName = user.lastName;
+
+  return `${firstName} ${lastName}`;
+}
+```
+
+🎉 Yep! 🎉
+
+```js
+// Good
+function getFullName(user) {
+  const { firstName, lastName } = user;
+  return `${firstName} ${lastName}`;
+}
+
+// Best
+function getFullName({ firstName, lastName }) {
+  return `${firstName} ${lastName}`;
+}
+```
+
+#### Resources
+
+- ESLint: [prefer-destructuring]
+- JSPerf: [Object Destructuring vs Not]
+
+### 4.2 Array Destructuring
+
+Avoid [destructuring arrays][Array Destructuring].
+
+> Why? As it turns out, performance goes down when destructuring arrays. 😞
+
+#### Examples
+
+🚫 Nope. 🚫
+
+```js
+const arr = [1, 2, 3, 4];
+const [first, second] = arr;
```

Comment:
I feel like sometimes array destructuring is more readable. Babel is able to transform this with `@babel/plugin-transform-destructuring` in cases where we are concerned about performance:

https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAhgJwTAvDA2gRgDQwEy4DMuALALoDcAUKJLOgGYCWC0uEAprQCa4B0AhB2hlU8JBSA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.0.0&envVersion=

Reply:
Oops, that link does not work. You have to enable the plugin by clicking "plugins" on the left, and searching for `@babel/plugin-transform-destructuring`

Reply:
> I feel like sometimes array destructuring is more readable.

@calebeby I definitely lean towards array destructuring. 

> Babel is able to transform this with @babel/plugin-transform-destructuring in cases where we are concerned about performance

Oh, this is a great point!

> You have to enable the plugin by clicking "plugins" on the left, and searching for `@babel/plugin-transform-destructuring`

I see this when I try to do this:
> Plugin @babel/plugin-transform-destructuring could not be loaded

Reply:
![screenshot_20181105-121114](https://user-images.githubusercontent.com/13206945/48023946-1beda000-e0f4-11e8-9022-aaafd1d7c8a9.png)

Sorry for the bad screenshot

Reply:
I've updated my PR based on this excellent feedback, thanks @calebeby! 😉